File: c/otterop/lang/result.c
#include <otterop/lang/int/result.h>
#include <gc.h>

typedef struct otterop_lang_Result_s {
    void *res;
    void *err;
} otterop_lang_Result_t;




otterop_lang_Result_t *otterop_lang_Result_new(void *res, void *err) {
    otterop_lang_Result_t *self = GC_malloc(sizeof(otterop_lang_Result_t));
    self->res = res;
    self->err = err;
    return self;
}

void *otterop_lang_Result_err(otterop_lang_Result_t *self) {
    return self->err;
}

void *otterop_lang_Result_unwrap(otterop_lang_Result_t *self) {
    return self->res;
}

otterop_lang_Result_t *otterop_lang_Result_of(void *res, void *err) {
    return otterop_lang_Result_new(res, err);
}

File: go/lang/result.go
package lang



type Result[RES any, ERR any] struct {
    res RES
    err ERR
}




func resultNew[RES any, ERR any](res RES, err ERR) *Result[RES, ERR] {
    this := new(Result[RES, ERR])
    this.res = res
    this.err = err
    return this
}

func (this *Result[RES, ERR]) Err() ERR {
    return this.err
}

func (this *Result[RES, ERR]) Unwrap() RES {
    return this.res
}

func Of[RES0 any, ERR0 any](res RES0, err ERR0) *Result[RES0, ERR0] {
    return resultNew[RES0, ERR0](res, err)
}